Add configurator.

goodov opened the pull request:
This PR adds base implementation for configuring reclient-enabled Chromium builds using existing Chromium configs.

goodov committed:
Add configurator.

ola-rozenfeld commented on README.md:
```diff
@@ -1,2 +1,138 @@
+Reclient-expected directory during the config write stage (see
+`ReclientCfg.rebase_if_path_value()`).
+
+### Cusomization via external py script
```
Typo: Customization

ola-rozenfeld commented on configure_reclient.py:
```diff
@@ -0,0 +1,564 @@
+        for directory in exec_root_included_dirs:
+            assert directory.startswith(
+                cls.exec_root
+            ), f'{directory} should be a part of {cls.exec_root}'
```
nit: "should be under" is a bit clearer than "a part of"

ola-rozenfeld commented on chromium-browser-clang/clang_remote_wrapper.template:
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
+# Copyright (c) 2023 The Brave Authors. All rights reserved.
```
Would the existing Apache license work? Something like:

```
# Copyright (c) 2023 The Brave Authors. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
```

ola-rozenfeld replied:
Actually, I'm not sure -- asking internally if we have a CLA or something in place. Will update.

ola-rozenfeld commented on README.md:
```diff
@@ -1,2 +1,138 @@
+support Chromium compilation on linux remotes via
+[Reclient](https://github.com/bazelbuild/reclient/).
+
+## Usage
```
I think it would be nice to include usage instructions specifically for building Chromium (and we'll add AOSP later) for various host/OS configurations.

For example:

## Building Chromium

### General

* Clone this repository (we refer to the path as `/path/to/reclient-configs/`).
* Optional: update this repository as part of your build `/*question to think about -- should we publish a sample wrapper script that clones/updates this repo to a given path, and then calls the configurator? Or is that something that every developer should decide whether they want to do?*/`

Add the following stanza to your `DEPS` file to run the configurator:

```
  {
    'name': 'configure_reclient',
    'pattern': '.',
    'action': ['python3', '/path/to/reclient-configs/configure_reclient.py', '--scr_dir=/path/to/chromium/src'],
  },
```

### Host OS: Linux, Target OS: Linux

Sample build:

```
$ gclient runhooks
$ gn gen out/reclient --args="use_remoteexec=true enable_nacl=false enable_rust=false"
$ autoninja -C out/reclient chrome
```

`/* and same for other host/target */`

ola-rozenfeld commented on README.md:
````diff
@@ -1,2 +1,138 @@
+# Called after the configuration has ended.
+def post_configure():
+    pass
+```
````
Since the script a required argument, I think we should list all arguments:

## Options

The script accepts the following parameters: `/*I didn't try to make sure the links work, but that's the idea*/`:

### --src_dir

**Required**. Absolute path to the Chromium checkout `src`.

### --exec_root

Reclient `exec_root` directory.

* Should match `rbe_exec_root` GN arg.
* Should be under [src_dir](#--src_dir).

**Default**: [src_dir](#--src_dir).

### --build_dir

Build directory. Used to calculate relative paths, can be left as default if you build in any out/* directory.

**Default**: [src_dir](#--src_dir)/out/a.

### --reclient_cfgs_dir

Path to Chromium reclient_cfgs directory.

**Default**: [src_dir](#--src_dir)/buildtools/reclient_cfgs.

### --clang_base_path

Chromium clang base path.

* Should match 'clang_base_path' GN arg.

**Default**: [src_dir](#--src_dir)/third_party/llvm-build/Release+Asserts

### --custom_py

Path to python script to customize generated reclient configs.

* See [Customization via external py script](#customization-via-external-py-script) for reference.
* See also [Brave usage example](future-link).

**Default**: None

### --force

Configure reclient even if the `RBE_service` environment variable is not set.

**Default**: False

goodov replied:
> Since the script a required argument, I think we should list all arguments

Only `--src_dir` is required argument. Everything else is optional. I'd like to avoid listing all args in the readme, because most of them are there to just support some exotic setups.

ola-rozenfeld commented on README.md:
```diff
@@ -1,2 +1,138 @@
+   `//buildtools/reclient_cfgs/reproxy_cfg_templates/reproxy.cfg.template` with
+   empty variables.
+2. Merge `reproxy.cfg`.
+3. Set auth-specific vars from `RBE_*` environment variables if they are set.
```
Why are we doing this? Wouldn't it be better to leave these as environment variables? I worry that this behavior is misleading in general -- people define/modify RBE_* variables for a reason, for things they specifically do not want to put in the configs.

If the only reason was to fold all environment variables into the appropriate configs to have one consistent source, then we should do that for all of them, not just the auth ones -- and that would be difficult to do because we don't know where each belongs (unless we want to run all reclient binaries with the `--help` option from inside the script :-)). But also that might harm consistency, not aid it (for people who are used to operate on environment variables).

goodov replied:
Agree. This may only be usable in our setup, so I'll just move this into our custom script.

ola-rozenfeld commented on configure_reclient.py:
```diff
@@ -0,0 +1,564 @@
+        assert isinstance(cfg, dict)
+        cfg_to_write = FileUtils.create_generated_header(
+            source_cfg_paths) + '\n\n'
+        for key, value in cfg.items():
```
How about we sort these? Otherwise, the configurator is not consistent.

```
        for key in sorted(cfg.keys()):
            formatted_value = cls.to_cfg_value(key, cfg[key])
```

goodov committed:
Sort cfg keys during writing.

goodov committed:
Should be a part of -> should be under.

goodov committed:
Remove RBE_ vars setting from environment into reproxy.cfg.

goodov committed:
Make more things configurable from a custom py script.

ola-rozenfeld commented:
Re: copyright: Ulf got back to me with this:

> We don't have a CLA. In this case, I think either MPL or Apache would be fine (I think Apache would be slightly better).
> Here's a writeup by someone with legal experience (#notlegaladvice): https://opensource.com/article/20/10/copyright-notices-open-source-software
> Here's a specific recommendation on what to do with copyright notices:
> https://www.linuxfoundation.org/blog/blog/copyright-notices-in-open-source-software-projects
> I'd probably go with something along the lines of "Copyright Contributors to the XYZ project."
> I would ask Aleksey to not use "The Brave Authors" given that the repo is in our org.

The first article basically explains why the whole copyright notice doesn't mean much these days, and the second article gives suggestions to what the copyright header should be, if we want one at all. For an example using that suggestion, see e.g. [Bazel code](https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/remote/RemoteRetrier.java#L1) which is a project with a lot of contributors, both individual and corporate.

So imo either omitting the header entirely in code files or going with something like this is fine:

```
# Copyright (c) 2023 Contributors to the reclient-configs project. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
```

ola-rozenfeld committed:
Support older Chromium tags

ola-rozenfeld commented:
I wanted to add https://github.com/EngFlow/reclient-configs/commit/853632286f7e05720a303e82a055388d0996416d , but couldn't figure out how to push it to your PR. This supports older Chromium tags (I tried on e.g. `108.0.5359.0`), before they had the directories defined as they are now. This is a safer way -- if there is an upstream directory, we use it, otherwise we have a reasonable default.
Feel free to make changes.
I also changed this:

```
    def generate_rewrapper_cfgs(self):
        for tool in ['chromium-browser-clang', 'python']:
            for platform in ['linux', 'mac', 'windows']:
                self.generate_rewrapper_cfg(tool, platform)
```

because it's simpler and will make it easier to support e.g. `nacl` later on.

goodov committed:
Small cleanup. os.path.isfile is fine to call directly.

goodov committed:
Replace copyrights.

goodov committed:
Add yapf formatter fix.

goodov commented:
> I wanted to add [8536322](https://github.com/EngFlow/reclient-configs/commit/853632286f7e05720a303e82a055388d0996416d) , but couldn't figure out how to push it to your PR. This supports older Chromium tags (I tried on e.g. `108.0.5359.0`), before they had the directories defined as they are now. This is a safer way -- if there is an upstream directory, we use it, otherwise we have a reasonable default. Feel free to make changes. I also changed this:
> 
> ```
>     def generate_rewrapper_cfgs(self):
>         for tool in ['chromium-browser-clang', 'python']:
>             for platform in ['linux', 'mac', 'windows']:
>                 self.generate_rewrapper_cfg(tool, platform)
> ```
> 
> because it's simpler and will make it easier to support e.g. `nacl` later on.

thanks! added into this PR (with a small fix).

goodov commented:
I think we can merge this in the current state and push other changes as separate PRs on top of the `main` branch.

This includes extending documentation, adding examples, etc.

goodov commented on configure_reclient.py:
```diff
@@ -0,0 +1,560 @@
+        # Load Chromium config for linux remote.
+        rewrapper_cfg_fname = f'linux/{tool}/rewrapper_linux.cfg'
+        rewrapper_cfg_file = f'{Paths.reclient_cfgs_dir}/{rewrapper_cfg_fname}'
+        if not os.path.isfile(rewrapper_cfg_file):
```
this may also fallback to old configs if Chromium decides to move configs around in `main`. We need to be aware of this scenario, because things might continue working, but the resulting configs will become more and more outdated.

This can also be solved with some `--fallback_pre_cr108` flag which will allow old builds, but won't have the downsides I described above.

Another way of doing this is to add `--base_reclient_cfgs_dir` which should default to `{reclient_cfgs_dir}`. We can put any custom base configs in a separate directory and use it when required.

ola-rozenfeld replied:
Yep, agreed!
